### PR TITLE
Refactor and update to new 1.x API

### PR DESCRIPTION
**Pull request by ahmadia**
- Refactored functionality out of Nutch into JobClient and Job classes
- Added NutchException
- Associate configurations with Nutch
- Associate crawls with JobClient
- Provide filtering of unrelated jobs in JobClient
- Move job shortcuts to JobClient
- Add ConfigClient and Config types
- Add SeedClient and Seed types
- Add tests for above types and basic functionality

Verified functionality:
- Job/JobClient interface
- Config interface
- Seed list injection
- Full-bore crawls

Unverified functionality:
- Admin interface

### CODE REVIEW

**Comment by ahmadia** [comment]
cc @chrismattmann.  I'll make better progress now that I've got a better handle on this code and its interactions with the server.

File: nutch/nutch.py [hunk 1]
```diff
@@ -182,63 +149,201 @@ def delete(self, id=None):
...
+        jobClient = self.Jobs(errorsFatal=True)
+
+        for step in crawlCycle:
+            (job, status) = jobClient.create(step, **args)
+            if status != 200:
+                die('Could not start %s on server %s, Aborting.' % (step, self.serverEndpoint))
+            time.sleep(1)
+            print(self.jobGetInfo(job.id)[0])
+            print('\nPress return to proceed to next step.\n', file=sys.stderr)
+            sys.stdin.read(1)
```
Comment (chrismattmann) [inline nutch/nutch.py]: I don't think we should make crawl interactive. It just should print status like in L317, and then move on.

**Comment by chrismattmann** [comment]
Overall great work. I think you can commit this, but I'd appreciate making crawl not interactive. Can you also issue the same PR to my repo, or I can do that too I'm trying to keep them in sync, but maintain a delineation for my class.

File: nutch/nutch.py (hunk 1 above)
Comment (ahmadia) [inline nutch/nutch.py]: I didn't touch that part of the code.  I agree that it shouldn't be interactive.

**Comment by ahmadia** [comment]
If you don't mind, I'll get a few more commits in so that it's more or less full-featured before I merge in.

**Comment by ahmadia** [comment]
Probably by the middle of next week.

**Comment by chrismattmann** [comment]
I'm going to go ahead and commit what you have here to my chrismattmann:master. Makes it easier to play with it and test.

**Comment by chrismattmann** [comment]
I made some fixes in https://github.com/memex-explorer/nutch-python/commit/0bc86d2f9c422abf112a398d25263871b376ccb5

**Comment by ahmadia** [comment]
Thanks!

**Comment by ahmadia** [comment]
Another refactor, this time I'm adding a low-level Server class to replace the `callServer` function.  I'm also trying to make the interface look more like `elasticsearch-py`, where possible.

**Comment by chrismattmann** [comment]
Quick thoughts - why keep refactoring? I mean it was functional right?

**Comment by ahmadia** [comment]
It was really hard to work with from Python, and there were a number of things that weren't working due to the changes in the REST API.

**Comment by chrismattmann** [comment]
Hmm, well I think these are two orthogonal things. If there were bugs we fix those bugs. But refactoring it to look like ESpy is aesthetic and I'm not getting it and unfortunately not supportive of doing that. I'd rather it look like tika-python and thus callServer I think is something we should keep around, etc. We can talk about this more at the hackathon.

**Comment by ahmadia** [comment]
A simple case is the output from listing `/jobs`.  In the current version, that returns a bunch of JSON and a status code.  In the refactored version, a non-200 status raises an exception and the response is converted into a list of Job objects, which in turn have methods associated with them.

If you want to keep `callServer` around, I can provide a similar interface that wraps the Server.call() function (or alternatively, pull callServer back out and have Server.call() wrap it).  My main motivations for providing a Server object is that I want to be able to pass information about how to call Nutch around between the various sub-objects, and I wanted to include bits of context like "raise errors on non-200 status codes" that could be easily set in the top-level Nutch object then forgotten about.

Some of this is aesthetic, and some of it has to do with encapsulating the JSON into usable code.  Let's chat more about this at the hackathon.

**Comment by ahmadia** [comment]
@chrismattmann - Are you still feeling unhappy with this?  I can pull the callServer code back out if it's the only barrier to merging.  Otherwise I can keep working on this branch and we can discuss at the hackathon.  I'm fine with merging in whatever functionality you like and discarding the rest.

File: nutch/__init__.py [hunk 2]
```diff
@@ -14,8 +14,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-try:
-    __import__('pkg_resources').declare_namespace(__name__)
-except ImportError:
-    from pkgutil import extend_path
-    __path__ = extend_path(__path__, __name__)
+from .nutch import Nutch, NutchException, Job, Config
```
Comment (ahmadia) [inline nutch/__init__.py]: note, deleting the above is mandatory as soon as we start making stuff accessible in the top-level nutch namespace.  namespace packages are a real PITA so it's better to just get rid of this now.

File: nutch/nutch.py [hunk 3]
```diff
@@ -56,189 +56,404 @@
...
+            echo2("Response status:", resp.status_code)
+        if resp.status_code != 200:
+            if self.raiseErrors:
+                error = NutchException("Unexpected server response: %d" % resp.status_code)
+                error.status_code = resp.status_code
+                raise error
+            else:
+                warn('Nutch server returned status:', resp.status_code)
+        content_type = resp.headers['content-type']
+        if content_type == 'application/json' and not forceText:
```
Comment (ahmadia) [inline nutch/nutch.py]: forceText is the workaround for the response not matching the headers (for now).

File: nutch/nutch.py [hunk 4]
```diff
@@ -56,189 +56,404 @@
...
+    def create(self, cid, configData):
+        """
+        Create a new named (cid) configuration from a parameter dictionary (config_data).
+        """
+
+        cid = self.server.call('post', "/config/%s" % cid, configData, TextAcceptHeader)
+        new_config = Config(cid, self.server)
+        return new_config
+
+    def __getitem__(self, item):
```
Comment (ahmadia) [inline nutch/nutch.py]: this provides nc.Configs()['default']

File: nutch/nutch.py [hunk 5]
```diff
@@ -56,189 +56,404 @@
...
+
+        # let's be optimistic...
+        config = Config(item, self.server)
+        if config.info():
+            return config
+
+        # not found!
+        raise KeyError
+
+    def __setitem__(self, key, value):
```
Comment (ahmadia) [inline nutch/nutch.py]: I couldn't get this working

File: nutch/nutch.py [hunk 6]
```diff
@@ -56,189 +56,404 @@
...
+        Create a job given a command
+        :param command: Nutch command, one of nutch.LegalJobs
+        :param args: Additional arguments to pass to the job
+        :return: The job id
+        """
+
         command = command.upper()
         if command not in LegalJobs:
-            warn('Nutch command must be one of: %s' % LegalJobs.join(', '))
+            warn('Nutch command must be one of: %s' % ', '.join(LegalJobs))
```
Comment (ahmadia) [inline nutch/nutch.py]: I think this was a bug.

File: nutch/test_nutch.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
+# encoding: utf-8
```
Comment (ahmadia) [inline nutch/test_nutch.py]: TESTS!

File: nutch/test_nutch.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Test Nutch API
+# Assumes a Nutch REST server is running on localhost
+# TODO: Package into Travis tests
```
Comment (ahmadia) [inline nutch/test_nutch.py]: With Travis, we can run PRs against a Travis configuration with a Nutch REST server installed.

File: nutch/test_nutch.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+import nutch
+import pytest
+import glob
+
```
Comment (ahmadia) [inline nutch/test_nutch.py]: All of the tests are interacting with the REST server, so these are more integration tests than unit tests.

File: nutch/test_nutch.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    cc = get_config_client()
+    assert cc
+
+def test_config_access():
+    cc = get_config_client()
+    default_config = cc['default']
+    # there has to be something smarter to check here
+    assert default_config.info()
+
+# I don't know how to get this working
```
Comment (ahmadia) [inline nutch/test_nutch.py]: This is the functionality I can't figure out how to get working.

File: nutch/test_nutch.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+def test_job_start():
+    jc = get_job_client()
+    old_jobs = jc.list()
+    inject_job = jc.inject()
+    updated_jobs = jc.list()
+    assert(len(updated_jobs) == len(old_jobs) + 1)
+
+    # awesome functionality for checking if this job is in a list of jobs
+    assert(inject_job not in old_jobs)
```
Comment (ahmadia) [inline nutch/test_nutch.py]: I think this is pretty sweet.

File: nutch/test_nutch.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+
+def test_job_client_lists():
+    # the default constructor uses a timestamp to create unique crawlIds
+    jc1 = get_job_client()
+    jc2 = get_job_client()
+
+    jc1_job = jc1.inject()
+
+    # only jobs with the same crawlId are returned in the list()
```
Comment (ahmadia) [inline nutch/test_nutch.py]: This is pretty sweet too.

File: nutch/test_nutch.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    assert(inject_job.info()['state'] == 'STOPPING')
+
+
+def test_job_abort():
+    jc = get_job_client()
+    inject_job = jc.inject()
+    inject_job.abort()
+    assert(inject_job.info()['state'] == 'KILLED')
+
+# How do we delete jobs using the REST API?  Is it even possible?
```
Comment (ahmadia) [inline nutch/test_nutch.py]: Is a delete method exposed?  I didn't see it anywhere in the documentation.

**Comment by ahmadia** [comment]
@chrismattmann - this is ready for review/merge.

**Comment by chrismattmann** [comment]
I love this. I am +1 to merge, and thus am merging it. You are awesome @ahmadia thank you.

**Pull request merged**